gpio/src/gpio.rs

```diff
@@ -0,0 +1,947 @@
+
+    use super::Error;
+    use super::*;
+
+    #[derive(Debug)]
+    pub(crate) struct DummyDevice {
+        ngpio: u16,
+        pub(crate) gpio_names: Vec<String>,
+        line_state: RwLock<Vec<LineState>>,
+        get_num_gpios_result: Result<u16>,
+        get_gpio_name_result: Result<String>,
+        get_direction_result: Result<u8>,
+        set_direction_result: Result<()>,
+        get_value_result: Result<u8>,
+        set_value_result: Result<()>,
+        set_irq_type_result: Result<()>,
+        pub(crate) wait_for_irq_result: Result<()>,
+    }
+
+    impl DummyDevice {
+        pub(crate) fn new(ngpio: u16) -> Self {
+            Self {
+                ngpio,
+                gpio_names: vec!['\0'.to_string(); ngpio.into()],
+                line_state: RwLock::new(vec![
+                    LineState {
+                        dir: VIRTIO_GPIO_DIRECTION_NONE,
+                        val: None,
+                        irq_type: VIRTIO_GPIO_IRQ_TYPE_NONE,
+                    };
+                    ngpio.into()
+                ]),
+                get_num_gpios_result: Ok(0),
+                get_gpio_name_result: Ok("".to_string()),
+                get_direction_result: Ok(0),
+                set_direction_result: Ok(()),
+                get_value_result: Ok(0),
+                set_value_result: Ok(()),
+                set_irq_type_result: Ok(()),
+                wait_for_irq_result: Ok(()),
+            }
+        }
+    }
+
+    impl GpioDevice for DummyDevice {
+        fn open(_device: u32) -> Result<Self>
+        where
+            Self: Sized,
+        {
+            Ok(DummyDevice::new(8))
+        }
+
+        fn get_num_gpios(&self) -> Result<u16> {
+            if self.get_num_gpios_result.is_err() {
+                return self.get_num_gpios_result;
+            }
+
+            Ok(self.ngpio)
+        }
+
+        fn get_gpio_name(&self, gpio: u16) -> Result<String> {
+            assert!((gpio as usize) < self.gpio_names.len());
+
+            if self.get_gpio_name_result.is_err() {
+                return self.get_gpio_name_result.clone();
+            }
+
+            Ok(self.gpio_names[gpio as usize].clone())
+        }
+
+        fn get_direction(&self, gpio: u16) -> Result<u8> {
+            if self.get_direction_result.is_err() {
+                return self.get_direction_result;
+            }
+
+            Ok(self.line_state.read().unwrap()[gpio as usize].dir)
+        }
+
+        fn set_direction(&self, gpio: u16, dir: u8, value: u32) -> Result<()> {
+            if self.set_direction_result.is_err() {
+                return self.set_direction_result;
+            }
+
+            self.line_state.write().unwrap()[gpio as usize].dir = dir;
+            self.line_state.write().unwrap()[gpio as usize].val = match dir as u8 {
+                VIRTIO_GPIO_DIRECTION_NONE => None,
+                VIRTIO_GPIO_DIRECTION_IN => self.line_state.read().unwrap()[gpio as usize].val,
```

Comment:
This is confusing...  Without all the fancy Rust veneers, this statement is equal to:

`self.line_state[gpio].val = self.line_state[gpio].val;`

Is this what you really wanted?

Reply:
I think, I was trying to avoid doing assignments inside individual cases of the `match` and wanted the block to simply returned what needs to be set. For Direction IN, we don't want value to be updated and so like this.

Reply:
As with everything in Rust, there is probably an idiomatic way of doing this but I don't know it either.  For the time being your solution will be just fine.